solidity/dashboard/src/connectors/ledger.js

```diff
@@ -0,0 +1,77 @@
+import { LedgerSubprovider } from '@0x/subproviders'
+import TransportU2F from '@ledgerhq/hw-transport-u2f'
+import AppEth from '@ledgerhq/hw-app-eth'
+import { AbstractHardwareWalletConnector } from './abstract'
+import { getChainIdFromV, getEthereumTxObj } from './utils'
+import web3Utils from 'web3-utils'
+import { getBufferFromHex } from '../utils/general.utils'
+
+export class LedgerProvider extends AbstractHardwareWalletConnector {
+  constructor(chainId) {
+    super(new CustomLedgerSubprovider(1101))
+  }
+}
+
+const ledgerEthereumClientFactoryAsync = async () => {
+  const ledgerConnection = await TransportU2F.create()
+  ledgerConnection.setExchangeTimeout(100000)
+  const ledgerEthClient = new AppEth(ledgerConnection)
+
+  return ledgerEthClient
+}
+
+class CustomLedgerSubprovider extends LedgerSubprovider {
+  chainId
+
+  constructor(chainId) {
+    super({ networkId: chainId, ledgerEthereumClientFactoryAsync, baseDerivationPath: '44\'/60\'' })
+    this.chainId = chainId
+  }
+
+  async signTransactionAsync(txData) {
+    LedgerSubprovider._validateTxParams(txData)
+    if (txData.from === undefined || !web3Utils.isAddress(txData.from)) {
+      throw new Error('Invalid address')
+    }
+    txData.chainId = this.chainId
+
+    const initialDerivedKeyInfo = await this._initialDerivedKeyInfoAsync()
+    const derivedKeyInfo = this._findDerivedKeyInfoForAddress(initialDerivedKeyInfo, txData.from)
+    const fullDerivationPath = derivedKeyInfo.derivationPath
+
+    try {
+      this._ledgerClientIfExists = await this._createLedgerClientAsync()
+      const tx = getEthereumTxObj(txData, this.chainId)
+
+      tx.raw[6] = this.chainId
+      tx.raw[7] = Buffer.from([])
+      tx.raw[8] = Buffer.from([])
+      const result = await this._ledgerClientIfExists.signTransaction(
+        fullDerivationPath,
+        tx.serialize().toString('hex')
+      )
+
+      let v = result.v
+      const rv = parseInt(v, 16)
+      let cv = this.chainId * 2 + 35
+      if (rv !== cv && (rv & cv) !== rv) {
```

Comment:
`(rv & cv) !== rv` - could you explain this to me?

I had something different, checking if the ledger `v` was even/odd and using that to add the sign bit.

Reply:
I have to refactor this snippet of code and add a comment, it's quite confusing. I'm going to apply [your approach](https://github.com/keep-network/tbtc-dapp/pull/205/files#diff-86a78d68537ecad0a90a50fc3100341dR60-R75).

So as you mentioned [here](https://github.com/keep-network/tbtc-dapp/pull/205/files#diff-86a78d68537ecad0a90a50fc3100341dR56) the transport layer only returns the lower 2 bytes. For `chainId = 1101` a transport layer returns => `v = be` because `2238 == 08be`:
* `rv = parseInt(v, 16) = 190`
*  `rv & cv => 190 & 2237 => 188` and `190 & 2238 => 190`
*  `(rv & cv) !== rv)` with this operation I check if have to add `1` according to the eip155 algorithm.